Feature/refresh log config

mtrupkin opened the pull request:
Add support for reloading logback configuration from centralized spring cloud config server for non-spring-boot applications.

Commit:
Allow null path info field returned from request

Commit:
Add support for reloading logback configuration from centralized spring cloud config server for non-spring-boot applications.

millerts commented on src/main/java/org/opentestsystem/delivery/logging/LoggingConfigRefresher.java:
```diff
@@ -0,0 +1,76 @@
+public class LoggingConfigRefresher {
+  private static final Logger logger = LoggerFactory.getLogger(LoggingConfigRefresher.class);
+
+  @Autowired
```
Please use constructor injection.  This doesn't look like it has to have field injection, and as a team we prefer constructor injection because it allows us to test and classes could be instantiated without spring if necessary.

mtrupkin replied:
I used the spring context reload event to detect app startup.  This made injecting beans unavailable at that time.  I will use an alternative app startup event in order to configure the object with a constructor.

millerts commented on src/main/java/org/opentestsystem/delivery/logging/EventParser.java:
```diff
@@ -35,7 +36,7 @@
   public Optional<EventInfo> parsePostHandle(final HttpServletRequest request, final HttpServletResponse response, final Object
     handler, EventLogger logger) {
-    return Optional.of(EventInfo.create(request.getPathInfo(), getEventDataFields(request)));
+    return Optional.of(EventInfo.create(StringUtils.trimToEmpty(request.getPathInfo()), getEventDataFields(request)));
```
I know this isn't part of this change, but why are we using Optional here?  It seems that things are never optional?  If the call to EventInfo.create can somehow return a null EventInfo then this should be changed to Optional.fromNullable() otherwise this will throw an NPE.

mtrupkin replied:
These EventParser methods were intended to be overridden. They are not required to return an event to log.

millerts commented on src/main/java/org/opentestsystem/delivery/logging/ConfigRefreshListener.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public class ConfigRefreshListener extends LoggingConfigRefresher implements MessageListener {
+  private static final Logger logger = LoggerFactory.getLogger(ConfigRefreshListener.class);
+  private final ObjectMapper objectMapper = new ObjectMapper();
```
I feel this should be injected via Spring (constructor not field).  I guess if it is null then you could create a new ObjectMapper but all the applications that will use this refresher already have an ObjectMapper configured.

millerts commented on src/main/java/org/opentestsystem/delivery/logging/ConfigRefreshListener.java:
```diff
@@ -0,0 +1,38 @@
+package org.opentestsystem.delivery.logging;
+
+
```
nit, extra line break

mtrupkin replied:
Agreed.

millerts commented on src/main/java/org/opentestsystem/delivery/logging/ConfigRefreshListener.java:
```diff
@@ -0,0 +1,38 @@
+ * When the message is received, the logback configuration is reloaded
+ * from the centralized spring cloud config server.
+ */
+public class ConfigRefreshListener extends LoggingConfigRefresher implements MessageListener {
```
I'm not sure why all these classes extend LoggingConfigRefreshser rather than having it injected.  It seems this message listener needs to refresh the logging not that it is a LoggingRefresher.

mtrupkin replied:
Agreed.  The problem came from the same comment as above.  I used the spring context reload event to detect app startup.  I'll try another app startup event.

millerts commented on src/main/java/org/opentestsystem/delivery/logging/LoggingConfigRefresher.java:
```diff
@@ -0,0 +1,76 @@
+  private Environment environment;
+
+  public void refreshLoggerConfig() {
+    final String configUrl = environment.getProperty("CONFIG_SERVICE_URL", "http://localhost:8888");
```
I assume this was for testing, but I don't think we should provide a default value.  Either it is configured or it is not.  I also wonder if this should just throw if not present.

millerts replied:
In addition, rather than getting the environment could you just inject the property via Spring?

mtrupkin replied:
Same comment as above for retrieving the config url value.  The default value seams reasonable, but removing will not effect the system very much.

millerts commented on src/main/java/org/opentestsystem/delivery/logging/LoggingConfigRefresher.java:
```diff
@@ -0,0 +1,76 @@
+
+  public void refreshLoggerConfig() {
+    final String configUrl = environment.getProperty("CONFIG_SERVICE_URL", "http://localhost:8888");
+    if (StringUtils.isNotBlank(configUrl)) {
```
Is it valid for configUrl to be blank?

Regardless, if this condition stays can we flip and early return? `if (StringUtils.isBlank(configUrl)) return;`.  It'll remove one layer of nesting and make it easier to read.

mtrupkin replied:
If the config url is blank, the traditional logback configuration mechanism is used.

mtrupkin replied:
I reversed the conditional logic.

millerts commented on src/main/java/org/opentestsystem/delivery/logging/LoggingConfigRefresher.java:
```diff
@@ -0,0 +1,76 @@
+            builder = UriComponentsBuilder.fromUriString(configUrl);
+          } else {
+            // add default http protocol if missing
+            builder = UriComponentsBuilder.fromHttpUrl(String.format("http://%s", configUrl));
```
I kinda feel like this is on the person configuring their application.  Is there a reason we are providing two ways to properly format the configUrl?  It is obvious to me that we must have it be served at http or https.  Seems like we should force the person using this class to configure it correctly

mtrupkin replied:
I removed suppling a protocol if it is missing.

millerts commented on src/main/java/org/opentestsystem/delivery/logging/LoggingConfigRefresher.java:
```diff
@@ -0,0 +1,76 @@
+          context.reset();
+          configurator.doConfigure(springConfigURL);
+        } catch (JoranException je) {
+          // StatusPrinter will handle this
```
I think there should be a log.error here.  Unless StatusPrinter is something we'll see locally if this starts failing.

mtrupkin replied:
I'm adding a warning log here, I believe the app can still function without proper logging configuration.

millerts commented on src/main/resources/rabbitmq-spring-cloud-context.xml:
```diff
@@ -0,0 +1,28 @@
+           http://www.springframework.org/schema/beans
+           http://www.springframework.org/schema/beans/spring-beans.xsd">
+
+    <rabbit:connection-factory id="connectionFactory" host="${rabbitmq.host}"
```
Does this have to be in XML?  Or can we use Spring annotations and a `@Configuration` type class?  Everything in this file.

wtritch replied:
I'd prefer a @Configuration class with a @ConditionalOnProperty(value = "rabbitmq.host") so that I don't need to have rabbit configured if I'm running projects locally.

wtritch commented on src/main/java/org/opentestsystem/delivery/logging/LoggingConfigRefresher.java:
```diff
@@ -0,0 +1,76 @@
+
+          final String profiles = StringUtils.join(environment.getActiveProfiles(), ",");
+
+          final String configLabel = environment.getProperty("CONFIG_LABEL", "master");
```
Rather than these environment.getProperty calls, I'd prefer to see a ConfigurationProperties POJO that gets autowired into this class.  Thoughts?

mtrupkin replied:
I used the spring context reload event to detect app startup. I'll try another app startup event and use java config.

Commit:
Use bean init instead of spring context refresh event.

Auto-wiring was not available when using the context refresh event.
Now using the spring bean init method to configure loading logging configuration from centralized configuration.

In addition, using spring java config vs xml config

Commit:
Merge branch 'develop' into feature/refresh-log-config

# Conflicts:
#	src/main/java/org/opentestsystem/delivery/logging/EventParser.java
